queryable-spec.ts

```diff
@@ -0,0 +1,421 @@
+  map: (bindings: Bindings, fn: (value: RDF.Term, key: RDF.Variable) => RDF.Term) => Bindings;
+  merge: (left: Bindings, right: Bindings) => Bindings;
+  mergeWith: (
+    merger: (left: RDF.Term, right: RDF.Term, key: RDF.Variable) => RDF.Term,
+    left: Bindings,
+    right: Bindings,
+  ) => Bindings;
+}
+
+/*
+ * Base interfaces to represent query results. These interfaces are generic
+ * with respect to the types of query metadata objects. These can be extended
+ * by implementors.
+ */
+
+interface QueryableResultMetadata<OrderItemsType extends TermName | RDF.Variable> {
+  cardinality?: number;
+  availableOrders?: QueryOperationOrder<OrderItemsType>[];
+  [key: string]: any;
+}
+
+interface BaseQueryableResult {
+  type: 'bindings' | 'quads' | 'boolean' | 'void';
+}
+
+interface QueryableResultBindings extends BaseQueryableResult {
+  type: 'bindings';
+  bindings(opts?: { order?: QueryOperationOrder<RDF.Variable> }): Stream<Bindings>;
+  variables: RDF.Variable[];
+  metadata(opts: { [key: string]: any }): Promise<QueryableResultMetadata<RDF.Variable>>;
+}
+    
+interface QueryableResultQuads extends BaseQueryableResult {
+  type: 'quads';
+  quads(opts?: { order?: QueryOperationOrder<TermName> }): Stream<RDF.Quad>;
+  metadata(opts: { [key: string]: any }): Promise<QueryableResultMetadata<TermName>>;
+}
+
+interface QueryableResultBoolean extends BaseQueryableResult {
+  type: 'boolean';
+  value(): Promise<boolean>;
+}
+
+interface QueryableResultVoid extends BaseQueryableResult {
+  type: 'void';
+  execute(): Promise<void>;
+}
+
+type QueryableResult = QueryableResultBindings | QueryableResultBoolean | QueryableResultQuads | QueryableResultVoid;
+
+/*
+ * Context objects provide a way to pass additional bits information to
+ * implementors, such as but not limited to:
+ * - data sources
+ * - base IRI for IRI resolution
+ * - timestamp for expression evaluation
+ * - query language
+ * - ...
+ */
+
+// SourceType can be anything the query engine defines
+// TODO: we may consider defining some standards, like 'string', RDF.Source, ...
+interface QueryableContext<SourceType> {
+  sources: [SourceType, ...SourceType[]];
+  queryTimestamp?: Date; // Required for certain SPARQL operations such as NOW().
+  [key: string]: any;
+}
+    
+interface QueryableStringContext<SourceType> extends QueryableContext<SourceType> {
+  queryFormat?: QueryableFormat; // defaults to { language: 'SPARQL', version: '1.1', extensions: [] }
+  baseIRI?: string; // Required for parsing SPARQL queries
+};
+
+interface QueryableAlgebraContext<SourceType> extends QueryableContext<SourceType> {};
+    
+interface QueryableFormat {
+  language: string; // Like 'SPARQL'
+  version: string; // Like '1.1'
+  extensions: string[]; // TODO: leave the syntax of these extensions open for now?
+}
+
+/**
+ * Placeholder to represent SPARQL Algebra trees.
+ */
+type Algebra = {};
```

Comment:
How are we intending to be express the dependency on sparqlalgebrajs in the spec?

Reply:
Binding the spec to `sparqlalgebrajs` doesn't seem right as the latter is a specific implementation. I think @rubensworks proposed we simply keep it as an empty object (as it is right now) for the time being and refer/link to Algebra typings in the doc? I'd be in favor of that.

Reply:
Actually, sorry, that would have to become `any` not to mess with implementations.

Reply:
+1 to `any` (with disclaimer that it will be spec'ed later)